// @yarnpkg/plugin-link/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkResolver = exports.LinkFetcher = exports.PortalResolver = exports.PortalFetcher = void 0;
const LinkFetcher_1 = require("./LinkFetcher");
Object.defineProperty(exports, "LinkFetcher", { enumerable: true, get: function () { return LinkFetcher_1.LinkFetcher; } });
const LinkResolver_1 = require("./LinkResolver");
Object.defineProperty(exports, "LinkResolver", { enumerable: true, get: function () { return LinkResolver_1.LinkResolver; } });
const PortalFetcher_1 = require("./PortalFetcher");
Object.defineProperty(exports, "PortalFetcher", { enumerable: true, get: function () { return PortalFetcher_1.PortalFetcher; } });
const PortalResolver_1 = require("./PortalResolver");
Object.defineProperty(exports, "PortalResolver", { enumerable: true, get: function () { return PortalResolver_1.PortalResolver; } });
const plugin = {
    fetchers: [
        LinkFetcher_1.LinkFetcher,
        PortalFetcher_1.PortalFetcher,
    ],
    resolvers: [
        LinkResolver_1.LinkResolver,
        PortalResolver_1.PortalResolver,
    ],
};
// eslint-disable-next-line arca/no-default-export
exports.default = plugin;
