// @yarnpkg/plugin-npm-cli/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpmWhoamiCommand = exports.NpmTagRemoveCommand = exports.NpmTagListCommand = exports.NpmTagAddCommand = exports.NpmPublishCommand = exports.NpmLogoutCommand = exports.NpmLoginCommand = exports.NpmInfoCommand = exports.NpmAuditCommand = exports.npmAuditUtils = exports.npmAuditTypes = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@yarnpkg/core");
const audit_1 = tslib_1.__importDefault(require("./commands/npm/audit"));
exports.NpmAuditCommand = audit_1.default;
const info_1 = tslib_1.__importDefault(require("./commands/npm/info"));
exports.NpmInfoCommand = info_1.default;
const login_1 = tslib_1.__importDefault(require("./commands/npm/login"));
exports.NpmLoginCommand = login_1.default;
const logout_1 = tslib_1.__importDefault(require("./commands/npm/logout"));
exports.NpmLogoutCommand = logout_1.default;
const publish_1 = tslib_1.__importDefault(require("./commands/npm/publish"));
exports.NpmPublishCommand = publish_1.default;
const add_1 = tslib_1.__importDefault(require("./commands/npm/tag/add"));
exports.NpmTagAddCommand = add_1.default;
const list_1 = tslib_1.__importDefault(require("./commands/npm/tag/list"));
exports.NpmTagListCommand = list_1.default;
const remove_1 = tslib_1.__importDefault(require("./commands/npm/tag/remove"));
exports.NpmTagRemoveCommand = remove_1.default;
const whoami_1 = tslib_1.__importDefault(require("./commands/npm/whoami"));
exports.NpmWhoamiCommand = whoami_1.default;
const npmAuditTypes = tslib_1.__importStar(require("./npmAuditTypes"));
exports.npmAuditTypes = npmAuditTypes;
const npmAuditUtils = tslib_1.__importStar(require("./npmAuditUtils"));
exports.npmAuditUtils = npmAuditUtils;
const plugin = {
    configuration: {
        npmPublishAccess: {
            description: `Default access of the published packages`,
            type: core_1.SettingsType.STRING,
            default: null,
        },
        npmAuditExcludePackages: {
            description: `Array of glob patterns of packages to exclude from npm audit`,
            type: core_1.SettingsType.STRING,
            default: [],
            isArray: true,
        },
        npmAuditIgnoreAdvisories: {
            description: `Array of glob patterns of advisory IDs to exclude from npm audit`,
            type: core_1.SettingsType.STRING,
            default: [],
            isArray: true,
        },
    },
    commands: [
        audit_1.default,
        info_1.default,
        login_1.default,
        logout_1.default,
        publish_1.default,
        add_1.default,
        list_1.default,
        remove_1.default,
        whoami_1.default,
    ],
};
// eslint-disable-next-line arca/no-default-export
exports.default = plugin;
