// @yarnpkg/plugin-init/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InitCommand = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@yarnpkg/core");
const init_1 = tslib_1.__importDefault(require("./commands/init"));
exports.InitCommand = init_1.default;
const plugin = {
    configuration: {
        initScope: {
            description: `Scope used when creating packages via the init command`,
            type: core_1.SettingsType.STRING,
            default: null,
        },
        initFields: {
            description: `Additional fields to set when creating packages via the init command`,
            type: core_1.SettingsType.MAP,
            valueDefinition: {
                description: ``,
                type: core_1.SettingsType.ANY,
            },
        },
        initEditorConfig: {
            description: `Extra rules to define in the generator editorconfig`,
            type: core_1.SettingsType.MAP,
            valueDefinition: {
                description: ``,
                type: core_1.SettingsType.ANY,
            },
        },
    },
    commands: [
        init_1.default,
    ],
};
// eslint-disable-next-line arca/no-default-export
exports.default = plugin;
