// @yarnpkg/plugin-dlx/lib/commands/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli_1 = require("@yarnpkg/cli");
const core_1 = require("@yarnpkg/core");
const clipanion_1 = require("clipanion");
// eslint-disable-next-line arca/no-default-export
class CreateCommand extends cli_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.pkg = clipanion_1.Option.String(`-p,--package`, {
            description: `The package to run the provided command from`,
        });
        this.quiet = clipanion_1.Option.Boolean(`-q,--quiet`, false, {
            description: `Only report critical errors instead of printing the full install logs`,
        });
        this.command = clipanion_1.Option.String();
        this.args = clipanion_1.Option.Proxy();
    }
    async execute() {
        const flags = [];
        if (this.pkg)
            flags.push(`--package`, this.pkg);
        if (this.quiet)
            flags.push(`--quiet`);
        // @foo -> @foo/create
        const command = this.command.replace(/^(@[^@/]+)(@|$)/, `$1/create$2`);
        const descriptor = core_1.structUtils.parseDescriptor(command);
        // @foo/app -> @foo/create-app
        // foo -> create-foo
        const modifiedIdent = !descriptor.name.match(/^create(-|$)/)
            ? descriptor.scope
                ? core_1.structUtils.makeIdent(descriptor.scope, `create-${descriptor.name}`)
                : core_1.structUtils.makeIdent(null, `create-${descriptor.name}`)
            : descriptor;
        let finalDescriptorString = core_1.structUtils.stringifyIdent(modifiedIdent);
        if (descriptor.range !== `unknown`)
            finalDescriptorString += `@${descriptor.range}`;
        return this.cli.run([`dlx`, ...flags, finalDescriptorString, ...this.args]);
    }
}
CreateCommand.paths = [
    [`create`],
];
exports.default = CreateCommand;
