// @yarnpkg/plugin-file/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TarballFileResolver = exports.TarballFileFetcher = exports.FileResolver = exports.FileFetcher = exports.fileUtils = void 0;
const tslib_1 = require("tslib");
const FileFetcher_1 = require("./FileFetcher");
Object.defineProperty(exports, "FileFetcher", { enumerable: true, get: function () { return FileFetcher_1.FileFetcher; } });
const FileResolver_1 = require("./FileResolver");
Object.defineProperty(exports, "FileResolver", { enumerable: true, get: function () { return FileResolver_1.FileResolver; } });
const TarballFileFetcher_1 = require("./TarballFileFetcher");
Object.defineProperty(exports, "TarballFileFetcher", { enumerable: true, get: function () { return TarballFileFetcher_1.TarballFileFetcher; } });
const TarballFileResolver_1 = require("./TarballFileResolver");
Object.defineProperty(exports, "TarballFileResolver", { enumerable: true, get: function () { return TarballFileResolver_1.TarballFileResolver; } });
const fileUtils = tslib_1.__importStar(require("./fileUtils"));
exports.fileUtils = fileUtils;
const plugin = {
    fetchers: [
        TarballFileFetcher_1.TarballFileFetcher,
        FileFetcher_1.FileFetcher,
    ],
    resolvers: [
        TarballFileResolver_1.TarballFileResolver,
        FileResolver_1.FileResolver,
    ],
};
// eslint-disable-next-line arca/no-default-export
exports.default = plugin;
