// @yarnpkg/plugin-version/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionUtils = exports.VersionCommand = exports.VersionCheckCommand = exports.VersionApplyCommand = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@yarnpkg/core");
const apply_1 = tslib_1.__importDefault(require("./commands/version/apply"));
exports.VersionApplyCommand = apply_1.default;
const check_1 = tslib_1.__importDefault(require("./commands/version/check"));
exports.VersionCheckCommand = check_1.default;
const version_1 = tslib_1.__importDefault(require("./commands/version"));
exports.VersionCommand = version_1.default;
const versionUtils = tslib_1.__importStar(require("./versionUtils"));
exports.versionUtils = versionUtils;
const plugin = {
    configuration: {
        deferredVersionFolder: {
            description: `Folder where are stored the versioning files`,
            type: core_1.SettingsType.ABSOLUTE_PATH,
            default: `./.yarn/versions`,
        },
        preferDeferredVersions: {
            description: `If true, running \`yarn version\` will assume the \`--deferred\` flag unless \`--immediate\` is set`,
            type: core_1.SettingsType.BOOLEAN,
            default: false,
        },
    },
    commands: [
        apply_1.default,
        check_1.default,
        version_1.default,
    ],
};
// eslint-disable-next-line arca/no-default-export
exports.default = plugin;
