// @yarnpkg/plugin-npm-cli/lib/commands/npm/audit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_1 = require("@yarnpkg/cli");
const core_1 = require("@yarnpkg/core");
const core_2 = require("@yarnpkg/core");
const plugin_npm_1 = require("@yarnpkg/plugin-npm");
const clipanion_1 = require("clipanion");
const micromatch_1 = tslib_1.__importDefault(require("micromatch"));
const t = tslib_1.__importStar(require("typanion"));
const npmAuditTypes = tslib_1.__importStar(require("../../npmAuditTypes"));
const npmAuditUtils = tslib_1.__importStar(require("../../npmAuditUtils"));
// eslint-disable-next-line arca/no-default-export
class NpmAuditCommand extends cli_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.all = clipanion_1.Option.Boolean(`-A,--all`, false, {
            description: `Audit dependencies from all workspaces`,
        });
        this.recursive = clipanion_1.Option.Boolean(`-R,--recursive`, false, {
            description: `Audit transitive dependencies as well`,
        });
        this.environment = clipanion_1.Option.String(`--environment`, npmAuditTypes.Environment.All, {
            description: `Which environments to cover`,
            validator: t.isEnum(npmAuditTypes.Environment),
        });
        this.json = clipanion_1.Option.Boolean(`--json`, false, {
            description: `Format the output as an NDJSON stream`,
        });
        this.noDeprecations = clipanion_1.Option.Boolean(`--no-deprecations`, false, {
            description: `Don't warn about deprecated packages`,
        });
        this.severity = clipanion_1.Option.String(`--severity`, npmAuditTypes.Severity.Info, {
            description: `Minimal severity requested for packages to be displayed`,
            validator: t.isEnum(npmAuditTypes.Severity),
        });
        this.excludes = clipanion_1.Option.Array(`--exclude`, [], {
            description: `Array of glob patterns of packages to exclude from audit`,
        });
        this.ignores = clipanion_1.Option.Array(`--ignore`, [], {
            description: `Array of glob patterns of advisory ID's to ignore in the audit report`,
        });
    }
    async execute() {
        const configuration = await core_1.Configuration.find(this.context.cwd, this.context.plugins);
        const { project, workspace } = await core_1.Project.find(configuration, this.context.cwd);
        if (!workspace)
            throw new cli_1.WorkspaceRequiredError(project.cwd, this.context.cwd);
        await project.restoreInstallState();
        const topLevelDependencies = npmAuditUtils.getTopLevelDependencies(project, workspace, { all: this.all, environment: this.environment });
        const packages = npmAuditUtils.getPackages(project, topLevelDependencies, { recursive: this.recursive });
        const excludedPackages = Array.from(new Set([
            ...configuration.get(`npmAuditExcludePackages`),
            ...this.excludes,
        ]));
        const payload = Object.create(null);
        for (const [packageName, versions] of packages)
            if (!excludedPackages.some(pattern => micromatch_1.default.isMatch(packageName, pattern)))
                payload[packageName] = [...versions.keys()];
        const registry = plugin_npm_1.npmConfigUtils.getAuditRegistry({ configuration });
        let result;
        const httpReport = await core_1.LightReport.start({
            configuration,
            stdout: this.context.stdout,
        }, async () => {
            // Note: No `await` there, so that the request is sent in parallel with the deprecation checks
            const auditRequest = plugin_npm_1.npmHttpUtils.post(`/-/npm/v1/security/advisories/bulk`, payload, {
                authType: plugin_npm_1.npmHttpUtils.AuthType.BEST_EFFORT,
                configuration,
                jsonResponse: true,
                registry,
            });
            const deprecations = this.noDeprecations ? [] : await Promise.all(Array.from(Object.entries(payload), async ([packageName, versions]) => {
                const registryData = await plugin_npm_1.npmHttpUtils.getPackageMetadata(core_2.structUtils.parseIdent(packageName), {
                    project,
                });
                return core_1.miscUtils.mapAndFilter(versions, version => {
                    const { deprecated } = registryData.versions[version];
                    // Apparently some packages have a `deprecated` field set to an empty string
                    // (even though that shouldn't be possible since `npm deprecate ... ""` undeprecates
                    // the package, completely removing the `deprecated` field). Both the npm website
                    // and all other package managers skip showing deprecation warnings in this case.
                    if (!deprecated)
                        return core_1.miscUtils.mapAndFilter.skip;
                    return [packageName, version, deprecated];
                });
            }));
            const auditResult = await auditRequest;
            for (const [packageName, version, message] of deprecations.flat(1)) {
                // No need to report the deprecation audit message if the package is already reported as vulnerable
                if (Object.hasOwn(auditResult, packageName))
                    if (auditResult[packageName].some(advisory => core_1.semverUtils.satisfiesWithPrereleases(version, advisory.vulnerable_versions)))
                        continue;
                auditResult[packageName] ??= [];
                auditResult[packageName].push({
                    id: `${packageName} (deprecation)`,
                    title: message.trim() || `This package has been deprecated.`,
                    severity: npmAuditTypes.Severity.Moderate,
                    vulnerable_versions: version,
                });
            }
            result = auditResult;
        });
        if (httpReport.hasErrors())
            return httpReport.exitCode();
        const severities = npmAuditUtils.getSeverityInclusions(this.severity);
        const ignoredAdvisories = Array.from(new Set([
            ...configuration.get(`npmAuditIgnoreAdvisories`),
            ...this.ignores,
        ]));
        const expandedResult = Object.create(null);
        for (const [packageName, advisories] of Object.entries(result)) {
            const filteredAdvisories = advisories.filter(advisory => {
                return !micromatch_1.default.isMatch(`${advisory.id}`, ignoredAdvisories) && severities.has(advisory.severity);
            });
            if (filteredAdvisories.length > 0) {
                expandedResult[packageName] = filteredAdvisories.map(advisory => {
                    const packageVersions = packages.get(packageName);
                    if (typeof packageVersions === `undefined`)
                        throw new Error(`Assertion failed: Expected the registry to only return packages that were requested`);
                    const versions = [...packageVersions.keys()].filter(version => {
                        return core_1.semverUtils.satisfiesWithPrereleases(version, advisory.vulnerable_versions);
                    });
                    const dependents = new Map();
                    for (const version of versions)
                        for (const pkg of packageVersions.get(version))
                            dependents.set(pkg.locatorHash, pkg);
                    return {
                        ...advisory,
                        versions,
                        dependents: [...dependents.values()],
                    };
                });
            }
        }
        const hasError = Object.keys(expandedResult).length > 0;
        if (hasError) {
            core_1.treeUtils.emitTree(npmAuditUtils.getReportTree(expandedResult), {
                configuration,
                json: this.json,
                stdout: this.context.stdout,
                separators: 2,
            });
            return 1;
        }
        await core_1.StreamReport.start({
            configuration,
            includeFooter: false,
            json: this.json,
            stdout: this.context.stdout,
        }, async (report) => {
            report.reportInfo(core_1.MessageName.EXCEPTION, `No audit suggestions`);
        });
        return hasError ? 1 : 0;
    }
}
NpmAuditCommand.paths = [
    [`npm`, `audit`],
];
NpmAuditCommand.usage = clipanion_1.Command.Usage({
    description: `perform a vulnerability audit against the installed packages`,
    details: `
      This command checks for known security reports on the packages you use. The reports are by default extracted from the npm registry, and may or may not be relevant to your actual program (not all vulnerabilities affect all code paths).

      For consistency with our other commands the default is to only check the direct dependencies for the active workspace. To extend this search to all workspaces, use \`-A,--all\`. To extend this search to both direct and transitive dependencies, use \`-R,--recursive\`.

      Applying the \`--severity\` flag will limit the audit table to vulnerabilities of the corresponding severity and above. Valid values are ${npmAuditUtils.allSeverities.map(value => `\`${value}\``).join(`, `)}.

      If the \`--json\` flag is set, Yarn will print the output exactly as received from the registry. Regardless of this flag, the process will exit with a non-zero exit code if a report is found for the selected packages.

      If certain packages produce false positives for a particular environment, the \`--exclude\` flag can be used to exclude any number of packages from the audit. This can also be set in the configuration file with the \`npmAuditExcludePackages\` option.

      If particular advisories are needed to be ignored, the \`--ignore\` flag can be used with Advisory ID's to ignore any number of advisories in the audit report. This can also be set in the configuration file with the \`npmAuditIgnoreAdvisories\` option.

      To understand the dependency tree requiring vulnerable packages, check the raw report with the \`--json\` flag or use \`yarn why package\` to get more information as to who depends on them.
    `,
    examples: [[
            `Checks for known security issues with the installed packages. The output is a list of known issues.`,
            `yarn npm audit`,
        ], [
            `Audit dependencies in all workspaces`,
            `yarn npm audit --all`,
        ], [
            `Limit auditing to \`dependencies\` (excludes \`devDependencies\`)`,
            `yarn npm audit --environment production`,
        ], [
            `Show audit report as valid JSON`,
            `yarn npm audit --json`,
        ], [
            `Audit all direct and transitive dependencies`,
            `yarn npm audit --recursive`,
        ], [
            `Output moderate (or more severe) vulnerabilities`,
            `yarn npm audit --severity moderate`,
        ], [
            `Exclude certain packages`,
            `yarn npm audit --exclude package1 --exclude package2`,
        ], [
            `Ignore specific advisories`,
            `yarn npm audit --ignore 1234567 --ignore 7654321`,
        ]],
});
exports.default = NpmAuditCommand;
