// @yarnpkg/plugin-essentials/lib/commands/explain/peerRequirements.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.explainPeerRequirements = void 0;
const tslib_1 = require("tslib");
const cli_1 = require("@yarnpkg/cli");
const core_1 = require("@yarnpkg/core");
const clipanion_1 = require("clipanion");
const t = tslib_1.__importStar(require("typanion"));
// eslint-disable-next-line arca/no-default-export
class ExplainPeerRequirementsCommand extends cli_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.hash = clipanion_1.Option.String({
            validator: t.cascade(t.isString(), [
                t.matchesRegExp(/^p[0-9a-f]{5}$/),
            ]),
        });
    }
    async execute() {
        const configuration = await core_1.Configuration.find(this.context.cwd, this.context.plugins);
        const { project } = await core_1.Project.find(configuration, this.context.cwd);
        await project.restoreInstallState({
            restoreResolutions: false,
        });
        await project.applyLightResolution();
        return await explainPeerRequirements(this.hash, project, {
            stdout: this.context.stdout,
        });
    }
}
ExplainPeerRequirementsCommand.paths = [
    [`explain`, `peer-requirements`],
];
ExplainPeerRequirementsCommand.usage = clipanion_1.Command.Usage({
    description: `explain a set of peer requirements`,
    details: `
      A set of peer requirements represents all peer requirements that a dependent must satisfy when providing a given peer request to a requester and its descendants.

      When the hash argument is specified, this command prints a detailed explanation of all requirements of the set corresponding to the hash and whether they're satisfied or not.

      When used without arguments, this command lists all sets of peer requirements and the corresponding hash that can be used to get detailed information about a given set.

      **Note:** A hash is a six-letter p-prefixed code that can be obtained from peer dependency warnings or from the list of all peer requirements (\`yarn explain peer-requirements\`).
    `,
    examples: [[
            `Explain the corresponding set of peer requirements for a hash`,
            `$0 explain peer-requirements p1a4ed`,
        ], [
            `List all sets of peer requirements`,
            `$0 explain peer-requirements`,
        ]],
});
exports.default = ExplainPeerRequirementsCommand;
async function explainPeerRequirements(peerRequirementsHash, project, opts) {
    const warning = project.peerWarnings.find(warning => {
        return warning.hash === peerRequirementsHash;
    });
    if (typeof warning === `undefined`)
        throw new Error(`No peerDependency requirements found for hash: "${peerRequirementsHash}"`);
    const report = await core_1.StreamReport.start({
        configuration: project.configuration,
        stdout: opts.stdout,
        includeFooter: false,
        includePrefix: false,
    }, async (report) => {
        const Marks = core_1.formatUtils.mark(project.configuration);
        switch (warning.type) {
            case core_1.PeerWarningType.NotCompatibleAggregate:
                {
                    report.reportInfo(core_1.MessageName.UNNAMED, `We have a problem with ${core_1.formatUtils.pretty(project.configuration, warning.requested, core_1.formatUtils.Type.IDENT)}, which is provided with version ${core_1.structUtils.prettyReference(project.configuration, warning.version)}.`);
                    report.reportInfo(core_1.MessageName.UNNAMED, `It is needed by the following direct dependencies of workspaces in your project:`);
                    report.reportSeparator();
                    for (const dependent of warning.requesters.values()) {
                        const dependentPkg = project.storedPackages.get(dependent.locatorHash);
                        if (!dependentPkg)
                            throw new Error(`Assertion failed: Expected the package to be registered`);
                        const descriptor = dependentPkg?.peerDependencies.get(warning.requested.identHash);
                        if (!descriptor)
                            throw new Error(`Assertion failed: Expected the package to list the peer dependency`);
                        const mark = core_1.semverUtils.satisfiesWithPrereleases(warning.version, descriptor.range)
                            ? Marks.Check
                            : Marks.Cross;
                        report.reportInfo(null, `  ${mark} ${core_1.structUtils.prettyLocator(project.configuration, dependent)} (via ${core_1.structUtils.prettyRange(project.configuration, descriptor.range)})`);
                    }
                    const transitiveLinks = [...warning.links.values()].filter(link => {
                        return !warning.requesters.has(link.locatorHash);
                    });
                    if (transitiveLinks.length > 0) {
                        report.reportSeparator();
                        report.reportInfo(core_1.MessageName.UNNAMED, `However, those packages themselves have more dependencies listing ${core_1.structUtils.prettyIdent(project.configuration, warning.requested)} as peer dependency:`);
                        report.reportSeparator();
                        for (const link of transitiveLinks) {
                            const linkPkg = project.storedPackages.get(link.locatorHash);
                            if (!linkPkg)
                                throw new Error(`Assertion failed: Expected the package to be registered`);
                            const descriptor = linkPkg?.peerDependencies.get(warning.requested.identHash);
                            if (!descriptor)
                                throw new Error(`Assertion failed: Expected the package to list the peer dependency`);
                            const mark = core_1.semverUtils.satisfiesWithPrereleases(warning.version, descriptor.range)
                                ? Marks.Check
                                : Marks.Cross;
                            report.reportInfo(null, `  ${mark} ${core_1.structUtils.prettyLocator(project.configuration, link)} (via ${core_1.structUtils.prettyRange(project.configuration, descriptor.range)})`);
                        }
                    }
                    const allRanges = Array.from(warning.links.values(), locator => {
                        const pkg = project.storedPackages.get(locator.locatorHash);
                        if (typeof pkg === `undefined`)
                            throw new Error(`Assertion failed: Expected the package to be registered`);
                        const peerDependency = pkg.peerDependencies.get(warning.requested.identHash);
                        if (typeof peerDependency === `undefined`)
                            throw new Error(`Assertion failed: Expected the ident to be registered`);
                        return peerDependency.range;
                    });
                    if (allRanges.length > 1) {
                        const resolvedRange = core_1.semverUtils.simplifyRanges(allRanges);
                        report.reportSeparator();
                        if (resolvedRange === null) {
                            report.reportInfo(core_1.MessageName.UNNAMED, `Unfortunately, put together, we found no single range that can satisfy all those peer requirements.`);
                            report.reportInfo(core_1.MessageName.UNNAMED, `Your best option may be to try to upgrade some dependencies with ${core_1.formatUtils.pretty(project.configuration, `yarn up`, core_1.formatUtils.Type.CODE)}, or silence the warning via ${core_1.formatUtils.pretty(project.configuration, `logFilters`, core_1.formatUtils.Type.CODE)}.`);
                        }
                        else {
                            report.reportInfo(core_1.MessageName.UNNAMED, `Put together, the final range we computed is ${core_1.formatUtils.pretty(project.configuration, resolvedRange, core_1.formatUtils.Type.RANGE)}`);
                        }
                    }
                }
                break;
            default:
                {
                    report.reportInfo(core_1.MessageName.UNNAMED, `The ${core_1.formatUtils.pretty(project.configuration, `yarn explain peer-requirements`, core_1.formatUtils.Type.CODE)} command doesn't support this warning type yet.`);
                }
                break;
        }
    });
    return report.exitCode();
}
exports.explainPeerRequirements = explainPeerRequirements;
